File: leetcode/stacks/20-valid-parentheses/ts/stack.ts
import { IStack } from "./stack.interface";

class Node<T> {
  data: T;
  next: Node<T> | null;
}

export class Stack<T> implements IStack<T> {
  private _size: number = 0;
  private _top: Node<T> | null = null;

  push(data: T): void {
    const node: Node<T> = {
      data: data,
      next: this._top,
    };

    this._top = node;
    this._size++;
  }

  pop(): T | null {
    if (this.isEmpty()) {
      return null;
    }

    if (this._top) {
      const oldNode: Node<T> = this._top;
      this._top = this._top.next;
      this._size--;
      return oldNode.data;
    }

    return null;
  }

  size(): number {
    return this._size;
  }

  isEmpty(): boolean {
    return this._size == 0;
  }
}

File: leetcode/stacks/20-valid-parentheses/c/stack.c
#include <stdlib.h>
#include "stack.h"

struct StackNode *stackCreate()
{
  struct StackNode *stack = calloc(1, sizeof(struct StackNode));
  stack->size = 0;
  stack->top = NULL;
  return stack;
}

void stackDelete(struct StackNode *stack)
{
  while (!IS_STACK_EMPTY(stack))
    stackPop(stack);

  free(stack);
}

void stackPush(struct StackNode *stack, void *item)
{
  struct Node *newTopNode = calloc(1, sizeof(struct Node));
  newTopNode->item = item;
  newTopNode->next = stack->top;

  stack->top = newTopNode;
  stack->size++;
}

void *stackPop(struct StackNode *stack)
{
  if (IS_STACK_EMPTY(stack))
    return NULL;

  struct Node *oldTopNode = stack->top;
  void *item = oldTopNode->item;

  stack->top = oldTopNode->next;
  stack->size--;
  free(oldTopNode);

  return item;
}
